Backend/services/notificationService.js

```diff
@@ -132,4 +161,177 @@ const updateImportantNotificationTypes = async () => {
+        centroids.forEach((centroid, idx) => {
+          const distance = calculateDistance(hour, centroid);
+          if (distance < minDistance) {
+            minDistance = distance;
+            closestCentroid = idx;
+          }
+        });
+        clusters[closestCentroid].push(hour);
+      });
+
+      // Step 5: Update centroids based on the mean of assigned points
+      const newCentroids = updateCentroids(clusters);
+
+      // Check for convergence if centroids do not change
+      if (JSON.stringify(newCentroids) === JSON.stringify(centroids)) break;
+
+      centroids = newCentroids;
+      iterations++;
+    }
+
+    // Step 6: Determine the best hour from the largest cluster
+    const bestCluster = clusters.reduce(
+      (maxCluster, cluster) =>
+        cluster.length > maxCluster.length ? cluster : maxCluster,
+      clusters[0]
+    );
+    const bestHour = Math.round(
+      bestCluster.reduce((acc, hour) => acc + hour, 0) / bestCluster.length
+    );
+    await prisma.userActivePeriods.upsert({
+      where: { userId: userId },
+      update: { mostActivePeriods: { set: [bestHour] } },
+      create: {
+        userId: userId,
+        mostActivePeriods: { set: [bestHour] },
+      },
+    });
+    return bestHour;
+  } catch (error) {
+    console.error("Error in K-means clustering:", error);
+  }
+};
+
+const scheduleNotifications = async (userId, newNotification, io) => {
+  const userSockets = Array.from(io.sockets.sockets.values()).filter(
+    (socket) => socket.userId == userId
+  );
+  // Add the new notification to the pending list
+  if (!pendingNotifications.has(userId)) {
+    pendingNotifications.set(userId, []);
+  }
+  pendingNotifications.get(userId).push(newNotification);
+
+  // Check if the job is already scheduled
+  if (scheduledJobs.has(userId)) {
+    return;
+  }
+
+  const bestHour = await performKMeansClustering(userId);
+  const now = new Date();
+
+  const testRunTime = new Date(now.getTime() + 10 * 1000);
+
+  const job = schedule.scheduleJob(testRunTime, async () => {
+    const notifications = pendingNotifications.get(userId) || [];
+    if (notifications.length > 0) {
+      if (userSockets.length > 0) {
+        notifications.forEach((notif) => {
+          userSockets[0].emit("notification", notif);
+        });
+      } else if (userSockets.length === 0) {
+        const user = await prisma.user.findUnique({
+          where: { id: userId },
+          select: { email: true },
+        });
+
+        if (user && user.email) {
+          await sendEmail(
+            user.email,
+            "Missed Notifications- World Collective",
+            `You have ${notifications.length} new notifications. Please log in to view them.`
+          );
+        }
+      }
+
+      await prisma.notification.updateMany({
+        where: { userId, isPending: true },
+        data: { isPending: false },
+      });
+
+      const pendingCount = await prisma.notification.count({
```

Comment:
shouldn't this be 0 after the previous line?

Reply:
yeah it would be. When the scheduled notification comes in, the pending count goes to 0. I was having a problem where when the scheduled notification comes in, the pending count didn't do that in real time because i was not emitting the pending count. Thats why I included that.